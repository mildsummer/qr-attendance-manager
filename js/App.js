import React, { Component } from "react";
import { ThemeProvider } from "react-native-elements";
import { SafeAreaProvider } from "react-native-safe-area-context";
import { Notifications } from "expo";
import { Alert, Platform } from "react-native";
import { Provider } from "react-redux";
import "./firebase";
import * as actions from "./actions";
import store from "./store";
import AppContainer from "./AppContainer";
import { setupAuthStateHandler } from "./setup";
import { addChangeDateListener } from "./utils/onChangeDate";

// setup
setupAuthStateHandler(store);
if (process.env.NODE_ENV !== "test") {
  addChangeDateListener(() => {
    console.log("date changed");
    store.dispatch({
      type: actions.CHANGE_DATE
    });
  });
}

fetch('https://qr-attendance-manager.firebaseapp.com/notification_category.json')
  .then((response) => response.json())
  .then(async (category) => {
    if (category.categoryId) {
      try {
        await Notifications.deleteCategoryAsync(category.categoryId);
        await Notifications.createCategoryAsync(category.categoryId, category.actions);
        console.log(category.categoryId, 'create success');
        Notifications.addListener(notification => {
          console.log('received', notification);
          const action = category.actions.find((action) => (action.actionId === notification.actionId));
          if (action) {
            Alert.alert(JSON.stringify(notification));
          }
        });
      } catch ({ error }) {
        Alert.alert(message);
      }
    }
  }).catch((error) => {
    console.log(error);
});

Notifications.addListener(notification => {
  store.dispatch({
    type: actions.RECEIVE_NOTIFICATION,
    payload: notification
  });
  if (notification.data.historyId) {
    if (notification.origin === "selected") {
      store.dispatch(
        actions.navigate({
          routeName: "HistoryDetail",
          params: { id: notification.data.historyId }
        })
      );
    } else if (notification.origin === "received") {
      Alert.alert(
        `[${notification.data.name}]コメントが届きました`,
        "詳細画面へ移動しますか？",
        [
          {
            text: "キャンセル",
            style: "cancel"
          },
          {
            text: "移動する",
            onPress: () => {
              store.dispatch(
                actions.navigate({
                  routeName: "HistoryDetail",
                  params: { id: notification.data.historyId }
                })
              );
            }
          }
        ]
      );
    }
  }
  if (typeof notification.data.badge === 'number' && Platform.OS === 'ios') {
    Notifications.setBadgeNumberAsync(notification.data.badge);
  }
  if (notification.data.local) {
    if (notification.data.localSchedulingOptions) {
      Notifications.scheduleLocalNotificationAsync(notification.data.local, notification.data.localSchedulingOptions);
    } else {
      Notifications.presentLocalNotificationAsync(notification.data.local);
    }
  }
});
Notifications.cancelAllScheduledNotificationsAsync();

export default class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <ThemeProvider>
          <SafeAreaProvider>
            <AppContainer />
          </SafeAreaProvider>
        </ThemeProvider>
      </Provider>
    );
  }
}
