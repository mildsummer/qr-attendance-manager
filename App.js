import React, { Component } from 'react';
import { Alert, TouchableOpacity } from "react-native";
import { ThemeProvider } from 'react-native-elements';
import { createAppContainer, NavigationActions, createSwitchNavigator } from 'react-navigation';
import { createStackNavigator } from 'react-navigation-stack';
import { createBottomTabNavigator } from 'react-navigation-tabs';
import { connect, Provider } from 'react-redux'
import Icon from "react-native-vector-icons/SimpleLineIcons";
import Iconions from "react-native-vector-icons/Ionicons";
import './js/utils/firebase';
import { store } from './js/redux'
import { auth } from "./js/utils/firebase";

// screens
import Loading from './js/screens/Loading';
import Login from './js/screens/Login';
import ResetPassword from './js/screens/ResetPassword';
import User from './js/screens/User';
import Reader from './js/screens/Reader';
import List from './js/screens/List';

const mapStateToProps = state => ({
  user: state.user.data,
  init: state.user.init
});

const AppNavigator = createAppContainer(createSwitchNavigator({
  Loading,
  SignedOut : {
    screen: createStackNavigator({
      Login: {
        screen: Login,
        navigationOptions: {
          title: 'ログイン',
          headerTintColor: '#fff',
          headerStyle: {
            borderBottomWidth: 0,
            backgroundColor: '#74dcd9',
          },
          headerTitleStyle: {
            fontWeight: 'bold'
          }
        }
      },
      ResetPassword: {
        screen: ResetPassword,
        navigationOptions: {
          title: 'パスワードをお忘れの方',
          headerTintColor: '#fff',
          headerStyle: {
            borderBottomWidth: 0,
            backgroundColor: '#74dcd9',
          },
          headerTitleStyle: {
            fontWeight: 'bold'
          }
        }
      }
    }, {
      initialRouteName: 'Login'
    })
  },
  SignedIn: {
    screen: createStackNavigator({
      User: {
        screen: createBottomTabNavigator({
          User: {
            screen: User,
            navigationOptions:  {
              title: 'ホーム',
              tabBarIcon: (<Icon name='user' size={17} />)
            }
          },
          Reader: {
            screen: Reader,
            navigationOptions:  {
              title: 'QR読み取り',
              tabBarIcon: (<Iconions name='ios-qr-scanner' size={17} />)
            }
          },
          List: {
            screen: List,
            navigationOptions:  {
              title: '履歴',
              tabBarIcon: (<Icon name='list' size={17} />)
            }
          }
        }, {
          initialRouteName: 'User',
          order: ['Reader', 'User', 'List'],
          tabBarOptions: {
            style: {
              borderTopWidth: 0
            }
          }
        }),
        navigationOptions: ({ navigation }) => ({
          title: ({
            User: 'ホーム',
            Reader: 'QR読み取り',
            List: '読み取り履歴'
          })[navigation.state.routes[navigation.state.index].key],
          headerTintColor: '#fff',
          headerStyle: {
            borderBottomWidth: 0,
            backgroundColor: '#74dcd9',
          },
          headerTitleStyle: {
            fontWeight: 'bold'
          },
          headerRight: (
            <TouchableOpacity
              style={{
                marginHorizontal: 6,
                padding: 10
              }}
              onPress={() => {
                auth.signOut()
                  .catch(({ message }) => {
                    Alert.alert(message);
                  });
              }}
            >
              <Icon
                name='logout'
                size={17}
                color='#fff'
              />
            </TouchableOpacity>
          )
        })
      }
    })
  }
}, {
  initialRouteName: 'Loading'
}));

const AppContainer = connect(mapStateToProps, {})(
  class extends Component {
    componentWillReceiveProps(nextProps) {
      const { user, init } = this.props;
      if (nextProps.init && !init && !nextProps.user) {
        this.navigate('Login');
      } else if (user && !nextProps.user) {
        this.navigate('Login');
      } else if (!user && nextProps.user) {
        this.navigate('User');
      }
    }

    navigate(routeName) {
      this.navigator && this.navigator.dispatch(NavigationActions.navigate({ routeName }));
    }

    render() {
      return (
        <AppNavigator
          ref={(ref) => {
            if (ref) {
              this.navigator = ref;
            }
          }}
        />
      );
    }
  }
);

export default class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <ThemeProvider>
          <AppContainer />
        </ThemeProvider>
      </Provider>
    );
  }
}
