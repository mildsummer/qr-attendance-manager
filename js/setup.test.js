import React from "react";
import renderer from "react-test-renderer";

import App from "./setup";

describe("<App />", () => {
  let tree = null;
  it("has 1 child", () => {
    tree = renderer.create(<App />).toJSON();
    expect(tree.children.length).toBe(1);
  });

  it("renders correctly", () => {
    expect(tree).toMatchSnapshot();
  });
});
