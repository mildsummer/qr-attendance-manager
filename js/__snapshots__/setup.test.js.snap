// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00c2ad",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#fff"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
